java/arcs/core/storage/referencemode/MessageQueue.kt

```diff
@@ -45,27 +47,47 @@ class MessageQueue(
     @ExperimentalCoroutinesApi
     suspend fun enqueue(message: Message): Boolean {
         require(message !is Message.Enqueued) { "Cannot enqueue an already-enqueued message." }
-        size.incrementAndGet().let {
-            if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-        }
         val deferred = CompletableDeferred<Boolean>(coroutineContext[Job.Key])
-        mutex.withLock { queue += message.toEnqueued(deferred) }
+        mutex.withLock { queue += QueueEntry.MessageEntry(message.toEnqueued(deferred)) }
         CoroutineScope(coroutineContext).launch { drainQueue() }
-        return deferred.await().also {
-            size.decrementAndGet().let {
-                if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-            }
-        }
+        return deferred.await()
     }
 
     private suspend fun drainQueue() {
         val messagesToProcess = mutex.withLock {
-            mutableListOf<Message>().also {
+            mutableListOf<QueueEntry>().also {
                 it.addAll(queue)
                 queue.clear()
             }
         }
-        messagesToProcess.forEach { processMessage(it); Unit }
+
+        messagesToProcess.forEach {
+            when (it) {
+                is QueueEntry.MessageEntry -> processMessage(it.message)
+                is QueueEntry.Flush -> Unit
+            }
+        }
+
+        mutex.withLock {
+            if (queue.isEmpty()) {
+                onMessageQueueEmpty()
+            }
+        }
+    }
+
+    /**
+     * Posts a message on the queue, and returns a [Job] this will be completed when the queue
+     * arrives at this message. This allows you to wait for all *current* pending messages to
+     * be processed.
+     */
+    suspend fun flush(): Job {
```

Comment:
I don't see why we need to wait for "current" pensing messages. Could we instead just provide an idle() method that completes when the queue is empty? that seems more natural to use from a refmodestore point of view.

Reply:
The only reason I needed that functionality was for testing; to wait for the flush Job to be processed before checking that the backing store size is 0.